### PR TITLE
chickenPackages_5.chicken: fix cross-compilation

### PR DESCRIPTION
**Pull request by Artturin**
###### Motivation for this change
Supersedes #104775

###### Things done
`nix build ".#pkgsCross.aarch64-multiplatform.chicken"`
`nix build ".#chicken"`

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by corngood — APPROVED** [review]
Looks good.  Thanks!

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 1]
```diff
@@ -19,43 +19,45 @@ stdenv.mkDerivation rec {
...
   makeFlags = [
-    "PLATFORM=${platform}" "PREFIX=$(out)"
+    "PLATFORM=${platform}"
+    "PREFIX=$(out)"
+    "HOSTSYSTEM=${hostPlatform.config}"
+    "TARGETSYSTEM=${targetPlatform.config}"
+    "LIBRARIAN=$(HOSTSYSTEM)-gcc-ar"
+    "C_COMPILER=$(CC)"
+    "CXX_COMPILER=$(CXX)"
   ] ++ (lib.optionals stdenv.isDarwin [
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chicken/5/chicken.nix:39]:
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [
```
we don't need those.

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 2]
```diff
@@ -19,43 +19,45 @@ stdenv.mkDerivation rec {
...
     "XCODE_TOOL_PATH=${darwin.binutils.bintools}/bin"
-    "C_COMPILER=$(CC)"
   ]);
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chicken/5/chicken.nix:41]:
```suggestion
  ];
```

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, makeWrapper, darwin, bootstrap-chicken ? null }:
+{ lib, stdenv, fetchurl, makeWrapper, targetPlatform, hostPlatform, darwin, bootstrap-chicken ? null }:
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/chicken/5/chicken.nix:1]: I would prefer if we wouldn't allow override to null here.

**Comment by Artturin** [comment]
chickenPackages.egg2nix fails with `> /nix/store/ffcrrgb6k5h7crr93dj90bi9fy1jvj16-binutils-2.35.1/bin/ld: cannot find crti.o: No such file or directory`

File: pkgs/development/compilers/chicken/5/chicken.nix (hunk 3 above)
Comment (Artturin) [reply pkgs/development/compilers/chicken/5/chicken.nix:1]: the code that calls this is 
```
chicken = callPackage ./chicken.nix {
  bootstrap-chicken = self.chicken.override { bootstrap-chicken = null; };
};
```
but removing the null `boostrap-chicken ? null` didn't  cause a rebuild so i guess its okay

**Comment by siraben** [comment]
`nix-build .#pkgsCross.raspberryPi.chicken` fails:

```ShellSession
$ nix log /nix/store/x503zkjmvxvhwz7y5dpd802gcnp6jyx6-chicken-5.2.0.drv | ix
http://ix.io/3Evw
```

**Comment by Mic92** [comment]
Did egg2nix already fail before? 
@GrahamcOfBorg build egg2nix

File: pkgs/development/compilers/chicken/5/chicken.nix [hunk 4]
```diff
@@ -19,43 +19,45 @@ stdenv.mkDerivation rec {
...
 
   makeFlags = [
-    "PLATFORM=${platform}" "PREFIX=$(out)"
-  ] ++ (lib.optionals stdenv.isDarwin [
-    "XCODE_TOOL_PATH=${darwin.binutils.bintools}/bin"
+    "PLATFORM=${platform}"
+    "PREFIX=$(out)"
+    "HOSTSYSTEM=${hostPlatform.config}"
+    "TARGETSYSTEM=${targetPlatform.config}"
+    "LIBRARIAN=$(HOSTSYSTEM)-gcc-ar"
```
Comment (Mindavi) [inline pkgs/development/compilers/chicken/5/chicken.nix:36]: Does this work without the `-gcc` part?

**Review by Mindavi — CHANGES_REQUESTED** [review]
Make sure that it also works with clang / compilers with a different prefix.

Otherwise, LGTM.

Tested building `chickenPackages_5.chicken` as native build and cross-compiled to `aarch64-multiplatform`.

Tested both native version and cross version by running `csi` (using qemu-aarch64 for the cross case). I don't know the language and haven't actually tried any expressions, but the interpreter seems to at least run.

File: pkgs/development/compilers/chicken/5/chicken.nix (hunk 4 above)
Comment (Artturin) [reply pkgs/development/compilers/chicken/5/chicken.nix]: yes

**Comment by siraben** [comment]
@Mindavi you can try `(define (fact n) (if (= n 0) 1 (* n (fact (- n 1)))))`

**Comment by Artturin** [comment]
`nix build ".#pkgsCross.aarch64-multiplatform.chickenPackages_5.egg2nix"`


```
chicken-matchable> /nix/store/9wxfxzd1bdfpirvv90nasik7ww10fyqw-stdenv-linux/setup: line 1345: chicken-install: command not found
```

if i do 
```
diff --git a/pkgs/development/compilers/chicken/5/eggDerivation.nix b/pkgs/development/compilers/chicken/5/eggDerivation.nix
index a6d19eaeb13..61a3c92de5a 100644
--- a/pkgs/development/compilers/chicken/5/eggDerivation.nix
+++ b/pkgs/development/compilers/chicken/5/eggDerivation.nix
@@ -17,8 +17,8 @@ in
 stdenv.mkDerivation ({
   name = "chicken-${name}";
   propagatedBuildInputs = buildInputs;
-  nativeBuildInputs = [ makeWrapper ];
-  buildInputs = [ chicken ];
+  nativeBuildInputs = [ makeWrapper chicken ];
+  buildInputs = [ ];
 
   CSC_OPTIONS = lib.concatStringsSep " " cscOptions;
 
```

then it fails with 
```
chicken-matchable> /nix/store/rc7ny3a5slwxzy3scgrmm8vg7m72birv-chicken-aarch64-unknown-linux-gnu-5.2.0/bin/chicken-install: line 3: /nix/store/rc7ny3a5slwxzy3scgrmm8vg7m72birv-chicken-aarch64-unknown-linux-gnu-5.2.0/bin/.chicken-install-wrapped: cannot execute binary file:Exec format error
```

**Pull request closed without merging**